clients/src/main/java/org/apache/kafka/common/utils/KafkaEventQueue.java

```diff
@@ -0,0 +1,390 @@
+                Long deadlineNs = deadlineNsCalculator.apply(existingDeadlineNs);
+                boolean queueWasEmpty = head.isSingleton();
+                boolean shouldSignal = false;
+                switch (eventContext.insertionType) {
+                    case APPEND:
+                        head.insertBefore(eventContext);
+                        if (queueWasEmpty) {
+                            shouldSignal = true;
+                        }
+                        break;
+                    case PREPEND:
+                        head.insertAfter(eventContext);
+                        if (queueWasEmpty) {
+                            shouldSignal = true;
+                        }
+                        break;
+                    case DEFERRED:
+                        if (deadlineNs == null) {
+                            throw new RuntimeException("Deferred event must have a deadline.");
+                        }
+                        break;
+                }
+                if (deadlineNs != null) {
+                    long insertNs =  deadlineNs;
+                    long prevStartNs = delayMap.isEmpty() ? Long.MAX_VALUE : delayMap.firstKey();
+                    // If the time in nanoseconds is already taken, take the next one.
+                    while (delayMap.putIfAbsent(insertNs, eventContext) != null) {
+                        insertNs++;
+                    }
+                    eventContext.deadlineNs = insertNs;
+                    // If the new timeout is before all the existing ones, wake up the
+                    // timeout thread.
+                    if (insertNs <= prevStartNs) {
+                        shouldSignal = true;
+                    }
+                }
+                if (shouldSignal) {
+                    cond.signal();
+                }
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    private final ReentrantLock lock;
+    private final Logger log;
+    private final Supplier<Throwable> closedExceptionSupplier;
+    private final EventHandler eventHandler;
+    private final Thread eventHandlerThread;
+    private long closingTimeNs;
+    private Event<?> cleanupEvent;
+
+    public KafkaEventQueue(LogContext logContext,
+                           String threadNamePrefix) {
+        this(logContext, threadNamePrefix, TimeoutException::new);
+    }
+
+    public KafkaEventQueue(LogContext logContext,
+                           String threadNamePrefix,
+                           Supplier<Throwable> closedExceptionSupplier) {
+        this.lock = new ReentrantLock();
+        this.log = logContext.logger(KafkaEventQueue.class);
+        this.closedExceptionSupplier = closedExceptionSupplier;
+        this.eventHandler = new EventHandler();
+        this.eventHandlerThread = new KafkaThread(threadNamePrefix + "EventHandler",
+            this.eventHandler, false);
+        this.closingTimeNs = Long.MAX_VALUE;
```

Comment:
I added JavaDoc to `closingTimeNs` to explain the special value.  I'm not sure an accessor function would help-- it might lead to confusion if someone forgot to check it and used the value directly, not knowing about the special value.